# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/JSScript.cpp
# Commit: 7a81a85d9881
# Full Hash: 7a81a85d988158e98b3fd5fb269059cd1bca5063
# Author: Yoshi Cheng-Hao Huang <allstars.chh@gmail.com>
# Date: 2020-06-17 14:49:45
# Regressor Bug: 1628204
# File Overlap Count: 2
# Description:
#   Bug 1628204 - Part 1: Move append to compressionFinishedList inside runTask. r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D78461
# ==============================================================================

diff -r e9de35da7a75 -r 7a81a85d9881 js/src/vm/JSScript.cpp
--- a/js/src/vm/JSScript.cpp	Wed Jun 17 06:54:07 2020 +0000
+++ b/js/src/vm/JSScript.cpp	Wed Jun 17 07:37:40 2020 +0000
@@ -2597,14 +2597,23 @@
 }
 
 void SourceCompressionTask::runTask() {
-  if (shouldCancel()) {
-    return;
-  }
-
-  ScriptSource* source = sourceHolder_.get();
-  MOZ_ASSERT(source->hasUncompressedSource());
-
-  source->performTaskWork(this);
+  if (!shouldCancel()) {
+    TraceLoggerThread* logger = TraceLoggerForCurrentThread();
+    AutoTraceLog logCompile(logger, TraceLogger_CompressSource);
+
+    ScriptSource* source = sourceHolder_.get();
+    MOZ_ASSERT(source->hasUncompressedSource());
+
+    source->performTaskWork(this);
+  }
+
+  {
+    AutoLockHelperThreadState lock;
+    AutoEnterOOMUnsafeRegion oomUnsafe;
+    if (!HelperThreadState().compressionFinishedList(lock).append(this)) {
+      oomUnsafe.crash("handleCompressionWorkload");
+    }
+  }
 }
 
 void ScriptSource::triggerConvertToCompressedSourceFromTask(
@@ -2671,7 +2680,7 @@
     // Attempt to compress.  This may not succeed if OOM happens, but (because
     // it ordinarily happens on a helper thread) no error will ever be set here.
     MOZ_ASSERT(!cx->isExceptionPending());
-    task->runTask();
+    ss->performTaskWork(task.get());
     MOZ_ASSERT(!cx->isExceptionPending());
 
     // Convert |ss| from uncompressed to compressed data.
